utils/sharded_map.go

```diff
@@ -0,0 +1,85 @@
+// Copyright 2020 Coinbase, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package utils
+
+import (
+	"github.com/segmentio/fasthash/fnv1a"
+)
+
+const (
+	// DefaultShards is the default number of shards
+	// to use in ShardedMap.
+	DefaultShards = 256
+)
+
+// shardMapEntry governs access to the shard of
+// the map contained at a particular index.
+type shardMapEntry struct {
+	mutex   *PriorityMutex
+	entries map[string]interface{}
+}
+
+// ShardedMap allows concurrent writes
+// to a map by sharding the map into some
+// number of independently locked subsections.
+type ShardedMap struct {
+	shards []*shardMapEntry
+}
+
+// NewShardedMap creates a new *ShardedMap
+// with some number of shards. The larger the
+// number provided for shards, the less lock
+// contention there will be.
+//
+// As a rule of thumb, shards should usually
+// be set to the concurrency of the caller.
+func NewShardedMap(shards int) *ShardedMap {
+	m := &ShardedMap{
+		shards: make([]*shardMapEntry, shards),
+	}
+
+	for i := 0; i < shards; i++ {
+		m.shards[i] = &shardMapEntry{
+			entries: map[string]interface{}{},
+			mutex:   new(PriorityMutex),
+		}
+	}
+
+	return m
+}
+
+// shardIndex returns the index of the shard
+// that could contain the key.
+func (m *ShardedMap) shardIndex(key string) int {
+	return int(fnv1a.HashString32(key) % uint32(len(m.shards)))
+}
+
+// Lock acquires the lock for a shard that could contain
+// the key. This syntax allows the caller to perform multiple
+// operations while holding the lock for a single shard.
+func (m *ShardedMap) Lock(key string, priority bool) map[string]interface{} {
```

Comment:
I observe that nothing about this package uses the resulting `map[string]interface{}` _as_ a map. This could conceivably be a more general `ShardedContainer` API, where `Lock` returns an `interface{}` and it's up to the caller to type-assert it to whatever kind of container they're sharding.

This means you'd also have to parameterize the constructor to construct caller-specified container shards rather than assuming they're maps.

One nice benefit of this approach is that (apart from the need to type-assert the result of `Lock`) it's more typesafe, in that the caller can say it's a `map[string]uint64` instead of having to settle for a `map[string]interface{}`.

Reply:
> I observe that nothing about this package uses the resulting map[string]interface{} as a map. This could conceivably be a more general ShardedContainer API, where Lock returns an interface{} and it's up to the caller to type-assert it to whatever kind of container they're sharding.

With the current API, the only thing that we would probably want to be asserted is that `interface{}` is a pointer. Otherwise, any modifications to the returned object would be ineffectual (or we return a pointer to what is stored, which may get hairy).

> One nice benefit of this approach is that (apart from the need to type-assert the result of Lock) it's more typesafe, in that the caller can say it's a map[string]uint64 instead of having to settle for a map[string]interface{}.

Because the caller (me in `MutexMap`) doesn't need to parse the value in the map if it doesn't exist, this API allows the caller to avoid unnecessary reflection. However, I see your point about why a `ShardedContainer` would be cool.

Reply:
I think this is a change we could make in a future PR once we have a clear use case in mind (likely just wrapping a `ShardedContainer` implementation). For now, I'd like to keep this optimized for the use case I have in mind (optimization of the `reconciler` and `storage` packages).